### PR TITLE
fixed buffer lapping

### PR DESCRIPTION
**Pull request by Dorokhov**

### CODE REVIEW

File: buffer.go [hunk 1]
```diff
@@ -8,6 +8,7 @@ import (
 	"syscall"
 
 	"github.com/ashishgandhi/buffer/binary"
+	"log"
```
Comment (twalwyn) [inline buffer.go]: Group std imports together at the top.

File: buffer.go [hunk 2]
```diff
@@ -384,7 +385,13 @@ func (b *Buffer) Read(data []byte, c Cursor) (n int, next Cursor, err error) {
 		return b.readFirst(data)
 	}
 
-	return b.readOffset(data, offset)
+	n1, c1, err := b.readOffset(data, offset)
```
Comment (twalwyn) [inline buffer.go]: There's not a need to panic here, I think. I would leave the code as it was before

File: buffer_test.go [hunk 3]
```diff
@@ -472,6 +472,40 @@ func mustInsert(t *testing.T, err error) {
...
+		beforeFirstRecord1   = []byte{1,2,3}
+		beforeFirstRecord2   = []byte{1,2,3,4,5,6,7,8,9}
+		firstRecord = []byte{5}
+		lastRecord = []byte{5,6}
+		data = make([]byte, 1<<10)
+	)
+
+	n := len(beforeFirstRecord2) + total
+	b, err := New(n, bufFile)
+
```
Comment (twalwyn) [inline buffer_test.go]: remove the newline here.

File: buffer_test.go [hunk 4]
```diff
@@ -472,6 +472,40 @@ func mustInsert(t *testing.T, err error) {
...
+	defer b.Close()
+
+	b.Insert(beforeFirstRecord1)
+	_,c,_ := b.ReadFirst(data)
+	b.Insert(beforeFirstRecord2)
+	b.Read(data,c)
+	b.Insert(firstRecord)
+	b.Insert(lastRecord)
+	_,_,err = b.Read(data,Cursor{7,3})
+
```
Comment (twalwyn) [inline buffer_test.go]: remove newline

**Review by twalwyn — CHANGES_REQUESTED** [review]
Is the purpose of this to show how the bug is reproduced?

File: buffer.go (hunk 2 above)
Comment (lwolf) [reply buffer.go]: maybe set `err` instead of panic and return it

Comment (kvap) [reply buffer.go]: @Dorokhov I wonder where the fix is. If it is the panic, then how do you test it?

Comment (Dorokhov) [reply buffer.go]: done

File: buffer_test.go (hunk 4 above)
Comment (Dorokhov) [reply buffer_test.go]: done

**Comment by lwolf** [comment]
I'm confused, what's the point of this review if your original PR is already merged into the upstream?

**Comment by Dorokhov** [comment]
I reverted the changes from the upstream so that diff displays both: test and the fix

**Comment by lwolf** [comment]
You reverted the master branch in your fork. Upstream is 
https://github.com/ashishgandhi/buffer/commits/master

File: buffer_test.go [hunk 5]
```diff
@@ -472,6 +472,38 @@ func mustInsert(t *testing.T, err error) {
...
+	b.Insert(beforeFirstRecord2)
+	b.Read(data,c)
+	b.Insert(firstRecord)
+	b.Insert(lastRecord)
+	_,_,err = b.Read(data,Cursor{7,3})
+	if err != nil {
+		t.Fatalf("b.Read err = %v; want read first record", err)
+	}
+
+	if !bytes.Equal(data[:len(firstRecord)], firstRecord) {
```
Comment (kvap) [inline buffer_test.go]: You compare a subslice of `data`, but the error message contains the whole `data`.

File: buffer_test.go [hunk 6]
```diff
@@ -472,6 +472,38 @@ func mustInsert(t *testing.T, err error) {
...
+	)
+
+	n := len(beforeFirstRecord2) + total
+	b, err := New(n, bufFile)
+	if err != nil {
+		t.Fatalf("Failed to create buffer: %v", err)
+	}
+	defer b.Close()
+
+	b.Insert(beforeFirstRecord1)
```
Comment (kvap) [inline buffer_test.go]: The test code is a bit hard to understand. The variable names are not obvious.

Comment (twalwyn) [reply buffer_test.go]: Maybe add a few comments to clarify what this is trying a achieve.

File: buffer_test.go [hunk 7]
```diff
@@ -472,6 +472,38 @@ func mustInsert(t *testing.T, err error) {
...
+	_,c,_ := b.ReadFirst(data)
```
Comment (twalwyn) [inline buffer_test.go]: It's more conventional to put spaces between the commas here.

File: buffer_test.go [hunk 8]
```diff
@@ -472,6 +472,38 @@ func mustInsert(t *testing.T, err error) {
...
+	b.Insert(beforeFirstRecord2)
+	b.Read(data,c)
```
Comment (twalwyn) [inline buffer_test.go]: Space between args here

File: buffer_test.go [hunk 9]
```diff
@@ -472,6 +472,38 @@ func mustInsert(t *testing.T, err error) {
...
+	b.Insert(firstRecord)
+	b.Insert(lastRecord)
+	_,_,err = b.Read(data,Cursor{7,3})
```
Comment (twalwyn) [inline buffer_test.go]: Space between commas here.

File: buffer.go [hunk 10]
```diff
@@ -380,7 +380,7 @@ func (b *Buffer) Read(data []byte, c Cursor) (n int, next Cursor, err error) {
 	}
 
 	f := b.frame(offset)
-	if f.size() == 0 || f.seq() != seq {
+	if f.size() == 0 || f.seq() != seq || offset < b.first {
```
Comment (bocharov) [inline buffer.go]: Comment explaining what's going on here would be great.

File: buffer_test.go [hunk 11]
```diff
@@ -472,6 +472,38 @@ func mustInsert(t *testing.T, err error) {
 	}
 }
 
+func TestBrokenCursorBeforeFirstRecord(t *testing.T) {
+	var (
+		beforeFirstRecord1   = []byte{1,2,3}
```
Comment (bocharov) [inline buffer_test.go]: It seems code is not "go fmt".

File: buffer_test.go [hunk 12]
```diff
@@ -472,6 +472,38 @@ func mustInsert(t *testing.T, err error) {
...
+		beforeFirstRecord2   = []byte{1,2,3,4,5,6,7,8,9}
```
Comment (bocharov) [inline buffer_test.go]: Having several test cases would be great - negative and positive.

File: buffer.go (hunk 10 above)
Comment (Dorokhov) [reply buffer.go]: done

File: buffer_test.go (hunk 11 above)
Comment (Dorokhov) [reply buffer_test.go]: done

File: buffer_test.go (hunk 12 above)
Comment (Dorokhov) [reply buffer_test.go]: tests for positive case exist, negative added

File: buffer_test.go (hunk 6 above)
Comment (Dorokhov) [reply buffer_test.go]: done

File: buffer_test.go (hunk 7 above)
Comment (Dorokhov) [reply buffer_test.go]: done

File: buffer_test.go (hunk 8 above)
Comment (Dorokhov) [reply buffer_test.go]: done

File: buffer_test.go (hunk 9 above)
Comment (Dorokhov) [reply buffer_test.go]: done

File: buffer_test.go (hunk 5 above)
Comment (Dorokhov) [reply buffer_test.go]: done

**Review by kvap — APPROVED** [review]
:100:

**Comment by twalwyn** [comment]
Cool!

**Review by twalwyn — APPROVED** [review]
🚢

**Pull request merged**